com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs

```diff
@@ -0,0 +1,110 @@
+
+        class CachedData
+        {
+            public PlayableGraph graph;
+            public AnimationPlayableOutput output;
+            public Dictionary<string, AnimationClipPlayable> playables = new Dictionary<string, AnimationClipPlayable>();
+            public PoseStateGroundTruthInfo poseState;
+        }
+        Dictionary<GameObject, CachedData> m_CacheMap;
+
+        /// <inheritdoc/>
+        protected override void OnCreate()
+        {
+            m_CacheMap = new Dictionary<GameObject, CachedData>();
+        }
+
+        CachedData InitializeCacheData(AnimationRandomizerTag tag)
+        {
+            var cachedData = new CachedData { graph = PlayableGraph.Create() };
+            cachedData.graph.SetTimeUpdateMode(DirectorUpdateMode.GameTime);
+            var animator = tag.gameObject.GetComponent<Animator>();
+            animator.applyRootMotion = tag.applyRootMotion;
+
+            cachedData.output = AnimationPlayableOutput.Create(cachedData.graph, "Animation", animator);
+
+            for (var i = 0; i < tag.animationClips.GetCategoryCount(); i++)
+            {
+                var clip = tag.animationClips.GetCategory(i);
+                var playable = AnimationClipPlayable.Create(cachedData.graph, clip.animationClip);
+                cachedData.playables[clip.animationClip.name] = playable;
+            }
+
+            cachedData.poseState = tag.gameObject.GetComponent<PoseStateGroundTruthInfo>();
+            if (cachedData.poseState == null)
+            {
+                cachedData.poseState = tag.gameObject.AddComponent<PoseStateGroundTruthInfo>();
+            }
+
+            return cachedData;
+        }
+
+        CachedData GetGraph(AnimationRandomizerTag tag)
+        {
+            if (!m_CacheMap.ContainsKey(tag.gameObject))
+            {
+                m_CacheMap[tag.gameObject] = InitializeCacheData(tag);
+
+            }
+            return m_CacheMap[tag.gameObject];
+        }
+
+        /// <inheritdoc/>
+        protected override void OnIterationStart()
+        {
+            var taggedObjects = tagManager.Query<AnimationRandomizerTag>();
+            foreach (var taggedObject in taggedObjects)
+            {
+                var tag = taggedObject.GetComponent<AnimationRandomizerTag>();
+                var cached = GetGraph(tag);
+
+                var clips = tag.animationClips;
+                var animationPoseLabel = clips.Sample();
+                var clip = animationPoseLabel.animationClip;
+
+                var playable = cached.playables[clip.name];
+                cached.output.SetSourcePlayable(cached.playables[clip.name]);
+                var t = m_FloatParameter.Sample() * clip.length;
+
+                playable.SetTime(t);
+                playable.SetSpeed(0);
+
+                cached.graph.Play();
+
+                cached.poseState.poseState = animationPoseLabel.GetPoseAtTime(t);
```

Comment:
Didn't we decide to separate the randomization from the labeling?

Reply:
Well, I wanted to go down that road but I talked to some folks on the animation team about trying to figure out the current clip and frame from the executing animation, and they really thought that was a bad idea. So I did go away from using the tag to store the info and just added a "cache" data item to the game object that had the labeling. But I definitely could change that so that we are just caching the current clip and time in the cache data, and put the labeling info back on to the labeler. That's probably the best idea.

Reply:
Ok, this has been reworked and I think it is more aligned to what you are looking for